### DISCUSSION TITLE
How can I add a new project to the current solution using Community.VisualStudio.Toolkit ?

### DISCUSSION THREAD

**Discussion by CplMulder** [VsixCommunity/Community.VisualStudio.Toolkit#361] [Q&A]
Hi there,

Thanks for the awesome toolkit...

I have an extension I wrote ages ago (still in VB haha) that I am porting to VS2019 and above.... I have almost everything working using the Community.VisualStudio.Toolkit but one thing I still cannot do that I used to be able to do with the DTE2 interface is to add a New project to the current solution (you used to point it to a template file)

Does anyone know if there any way to do this using the toolkit?

**Comment by ErikEJ** [comment]
What did you do before?

**Reply by CplMulder** [reply]
Using the EnvDTE/EnvDTE80 namespace and the DTE2 object had Solution2 property for the current solution on which you could call **.AddFromTemplate(_templatepath, projectpath, projectname_)**

**Reply by CplMulder** [reply]
... and you used to get the path by calling something like **Solution2.GetProjectTemplate(String.Format("{0}.zip", projectType), projectLanguage)**

**Comment by reduckted** [comment]
Based on how a solution folder is added here:
https://github.com/VsixCommunity/Community.VisualStudio.Toolkit/blob/9c32305c83827d0bfd7c85777f9e79fc71e22c2b/src/toolkit/Community.VisualStudio.Toolkit.Shared/Solution/Solution.cs#L30-L34

I suspect that at least part of the answer is to use the [`IVsSolution.CreateProject`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.shell.interop.ivssolution.createproject?view=visualstudiosdk-2022) method.

Based on that documentation, I _think_ you'd want to call it something like this to create a project from a template:

```cs
Guid guidNull = default;
Guid iidProject = typeof(IVsHierarchy).GUID;

solution.CreateProject(
    rguidProjectType: ref guidNull,
    lpszMoniker: pathToTemplate, 
    lpszLocation: pathToNewProjectDirectory, 
    lpszName: newProjectName, 
    grfCreateFlags: __VSCREATEPROJFLAGS.CPF_CLONEFILE,
    iidProject: typeof(IVsHierarchy).GUID,
    ppProject: out IntPtr projectPtr
);
```

Where you're supposed to get the project template path from, I have no idea. 🤷‍♂️ 

There's also seems to be a few other methods to add projects that you might have some luck with (I haven't tried any of them though):
* [`IVsSolution2.CreateProject`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.shell.interop.ivssolution2.createproject?view=visualstudiosdk-2022) (looks identical to `IVsSolution.CreateProject`)
* [`IVsSolution3.CreateNewProjectViaDlgEx`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.shell.interop.ivssolution3.createnewprojectviadlgex?view=visualstudiosdk-2022)
* [`IVsSolution6.AddNewProjectFromTemplate`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.shell.interop.ivssolution6.addnewprojectfromtemplate?view=visualstudiosdk-2022)
* [`IVsSolution6.AddExistingProject`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.shell.interop.ivssolution6.addexistingproject?view=visualstudiosdk-2022)

**Reply by CplMulder** [reply]
Thanks @reduckted - I will hack about with those and see what I come up with... appreciated.
